test/powershell/Modules/Microsoft.PowerShell.Core/Job.Tests.ps1

```diff
@@ -22,28 +35,119 @@ Describe "Job Cmdlet Tests" -Tag "CI" {
             $waitjob = Wait-Job -Timeout 10 -id $j.id
             if ( $waitJob ) {
                 $result = receive-job -id $j.id
-                $result.id | Should be 2
+                $result | Should be 2
             }
         }
     }
     Context "jobs which take time" {
+        BeforeAll {
+            $jobBlock =
+            {
+                $count = 5
+
+                while ($count)
+                {
+                    Write-Output "count: $count"
+                    --$count
+                    start-sleep 2
+                }
+            }
+        }
         BeforeEach {
-            $j = start-job -scriptblock { Start-Sleep 15 }
+            $j = start-job -scriptblock $jobBlock
         }
         AfterEach {
             Get-Job | Remove-Job -force
         }
         It "Wait-Job will wait for a job" {
             $result = wait-Job $j
             $result | should be $j
+            $j.State | should be Completed
         }
-        It "Stop-Job will stop a job" -pending {
+        It "Wait-Job will timeout waiting for a job" {
+            $result = Wait-Job -Timeout 2 $j
+            $result | should be $null
+        }
+        It "Stop-Job will stop a job" {
             Stop-Job -id $j.id
-            $j.Status |Should be Stopped
+            $j.State | should be Stopped
+        }
+        It "Remove-Job will not remove a running job" {
+            $id = $j.Id
+            $j | remove-job -ErrorAction SilentlyContinue
+            $job = Get-Job -Id $id
+            $job | should be $j
+        }
+        It "Remove-Job -Force will remove a running job" {
+            $id = $j.Id
+            $j | remove-job -ErrorAction SilentlyContinue
+            $job = Get-Job -Id $id
+            $job | should be $j
+            $j | Remove-Job -Force
+            $job = Get-Job -Id $id -ErrorAction SilentlyContinue
+            $job | should be $null
+        }
+    }
+    Context "Retrieving partial output from jobs" {
+        BeforeAll {
+            function getResults($job, $n, $keep)
+            {
+                $results = @()
+
+                while ($results.Count -lt $n)
+                {
+                    if ($keep)
+                    {
+                        $results = @() + (Receive-Job -Keep $job)
+                    }
+                    else
+                    {
+                        $results += Receive-Job $job
+                    }
+                }
+
+                return $results
+            }
+
+            # call this only within an It block
+            function checkContent($ar)
+            {
+                $v = $ar[0] - 1
+
+                for ($i=0; $i -lt $ar.Count; ++$i)
+                {
+                    $ar[$i] | should be ($v + 1)
+                    $v = $ar[$i]
+                }
+            }
+
+        }
+        BeforeEach {
+            $j = start-job -scriptblock { $count = 1; while ($true) { Write-Output ($count++) } }
```

Comment:
I think @iSazonov means this:
```
BeforeAll {
    New-Item $TestDrive\jobs.txt -type File
    $j = Start-Job { get-content $TestDrive\jobs.txt -wait }
}

AfterAll {
    Get-Job | Remove-Job -Force
}

It "....." {
     Add-Content $TestDrive\job.txt -Value "Hello"
     Receive-Job $j | should be "Hello"
     Add-Content $TestDrive\job.txt -Value "World"
     Receive-Job $j | should be "World"
}

It "...." {
    Add-Content $TestDrive\job.txt -Value "Keep"
    Receive-Job $j -Keep | should be "Keep"
    Add-Content $TestDrive\job.txt -Value "Output"
    (Receive-Job $j) -Join "," | should be "Keep,Output"
}
```
I think this is better than the loop approach because 1) it's simpler to read and understand; 2) in case that `receive-job` really is broken, the loop might leave the test in a bad place.